packages/wrangler/src/d1/execute.tsx

```diff
@@ -348,41 +346,194 @@ async function executeRemotely({
-				},
-				body: JSON.stringify({ sql }),
+				results: [
+					{
+						"Total queries executed": numQueries,
+						"Rows read": meta.rows_read,
+						"Rows written": meta.rows_written,
+						"Databas size (MB)": (meta.size_after / 1_000_000).toFixed(2),
+					},
+				],
+				success: true,
+				finalBookmark,
+				meta,
+			},
+		];
+	} else {
+		const result = await d1ApiPost<QueryResult[]>(accountId, db, "query", {
+			sql: input.command,
+		});
+		logResult(result);
+		return result;
+	}
+}
+
+async function uploadAndBeginIngestion(
+	accountId: string,
+	db: Database,
+	file: string,
+	etag: string,
+	initResponse: ImportInitResponse
+) {
+	const { uploadUrl, filename } = initResponse;
+	logger.log(`🌀 Uploading ${filename}...`);
+
+	const { size } = await fs.stat(file);
+
+	const uploadResponse = await fetch(uploadUrl, {
+		method: "PUT",
+		headers: {
+			"Content-length": `${size}`,
+		},
+		body: createReadStream(file),
+		duplex: "half", // required for NodeJS streams over .fetch ?
+	});
+
+	if (uploadResponse.status !== 200) {
+		throw new UserError(
+			`File could not be uploaded. Please retry.\nGot response: ${await uploadResponse.text()}`
+		);
+	}
+
+	const etagResponse = uploadResponse.headers.get("etag");
+	if (!etagResponse) {
+		throw new UserError(`File did not upload successfully. Please retry.`);
+	}
+	if (etag !== etagResponse.replace(/^"|"$/g, "")) {
+		throw new UserError(
+			`File contents did not upload successfully. Please retry.`
+		);
+	}
+	logger.log(`🌀 Uploading complete.`);
+
+	return await d1ApiPost<ImportPollingResponse | PollingFailure>(
+		accountId,
+		db,
+		"import",
+		{ action: "ingest", filename, etag }
+	);
+}
+
+async function pollUntilComplete(
+	response: ImportPollingResponse | PollingFailure,
+	accountId: string,
+	db: Database
+): Promise<ImportPollingResponse> {
+	if (!response.success) throw new Error(response.error);
+
+	response.messages.forEach((line) => {
+		logger.log(`🌀 ${line}`);
+	});
+
+	if (response.status === "complete") {
+		return response;
+	} else if (response.status === "error") {
+		throw new APIError({
+			text: response.errors?.join("\n"),
+			notes: response.messages.map((text) => ({ text })),
+		});
+	} else {
+		const newResponse = await d1ApiPost<ImportPollingResponse | PollingFailure>(
+			accountId,
+			db,
+			"import",
+			{
+				action: "poll",
+				currentBookmark: response.at_bookmark,
 			}
 		);
-		logResult(result);
-		results.push(...result);
+		return await pollUntilComplete(newResponse, accountId, db);
 	}
-	return results;
+}
+
+async function d1ApiPost<T>(
+	accountId: string,
+	db: Database,
+	action: string,
+	body: unknown
+) {
+	return await fetchResult<T>(
+		`/accounts/${accountId}/d1/database/${db.uuid}/${action}`,
+		{
+			method: "POST",
+			headers: {
+				"Content-Type": "application/json",
+				...(db.internal_env ? { "x-d1-internal-env": db.internal_env } : {}),
+			},
+			body: JSON.stringify(body),
+		}
+	);
```

Comment:
Is this needed? We already have utility methods for calling the Cloudflare API

Reply:
Isn't that what `fetchResult` is? I just extracted this so that all staging DB calls sent the right header, but if there's another internal api I should use for doing the fetch then lmk